File: Determinastic_growth/VFI/VFI.m
clear all clc
tic;

% Parameterization
alpha = 0.3333333333; delta = 0.1; beta = 0.975; A = 1.05;
nkgrid = 1001; maxDiff = 10.0; tol = 1.0e-10; iter = 0;

% Find the Steady-states
kss   = ((1 + beta * (delta - 1))/(alpha * beta * A))^(1 / (alpha - 1));
yss   = A * kss^alpha + (1 - delta)*kss;
css   = yss - kss;

% Construct Capital Grid
kmin = 0.5*kss; kmax = 1.5*kss;
k_grid = linspace(kmin, kmax, nkgrid);
kp_grid = linspace(kmin, kmax, nkgrid);

%rtrn = zeros(ngrid_k, ngrid_k);
V_old = zeros(1, nkgrid);
V_new = zeros(1, nkgrid);
k_pol = zeros(1, nkgrid);

% Main loop
while maxDiff > tol
    iter = iter + 1;
    fprintf('Current Iteration is: %.d \n', iter);
    
    temp = zeros(nkgrid, nkgrid);
    
    % Clear negative consumption first
    for i = 1:nkgrid
        for j = 1:nkgrid
          rtrn = A*(k_grid(i)^alpha) + (1-delta)*k_grid(i) - kp_grid(j);
          if rtrn < 0
            temp(i, j) = -100.0;
          else
            temp(i, j) = log(rtrn) + beta*V_old(j);
          end
        end
    end
    
    % Find Max and argmax
    [argvalue, argmax]= max(temp,[],2);
    V_new = argvalue;
    
   for i = 1:nkgrid
        k_pol(i) = k_grid(argmax(i));
   end
   
    % Apply MacQueen-Porteus Bound in every iteration
   c_low = (beta/(1-beta))*min(V_new-V_old);
   c_high = (beta/(1-beta))*max(V_new-V_old);
   maxDiff = c_high-c_low;
    
   %maxDiff = norm(V_new-V_old);
   V_old = V_new;  
end
toc; 

V_new = V_new + ((c_high+c_low)/2);


figure(1)
plot(k_grid, V_new, 'LineWidth',2, 'DisplayName','Value Function');
legend(); grid on;
xlim([min(k_grid) max(k_grid)])
title('Value Function Converged')
xlabel('level of Capital Today')
ylabel('Value')

figure(2)
plot(k_grid, k_pol, 'LineWidth',2, 'DisplayName','Capital Policy');
legend(); grid on;
xlim([min(k_grid) max(k_grid)])
title('Optimal Capital Policy Function')
xlabel('level of Capital Today')
ylabel('level of Capital Tomorrow')

File: 1_Deterministic_growth/VFI.f90
SUBROUTINE linspace(d1,d2,n,grid)
    IMPLICIT NONE

    INTEGER, INTENT(IN)                         :: n
    DOUBLE PRECISION, INTENT(IN)                :: d1, d2
    DOUBLE PRECISION, DIMENSION(n), INTENT(OUT) :: grid
    INTEGER :: idx

    grid(1) = d1
    DO idx = 0,n-2
       grid(idx+1) = d1+(DBLE(idx)*(d2-d1))/DBLE(n-1)
    END DO
    grid(n) = d2

END SUBROUTINE



PROGRAM VFI_determ
  IMPLICIT NONE
  external linspace
  !GRID SIZES
  INTEGER, parameter :: nkgrid  = 101
  INTEGER  :: iter = 0
  INTEGER  :: i, j

  !PARAMETERS
  REAL(8), parameter   :: alpha = 0.33333333333
  REAL(8), parameter   :: beta = 0.975
  REAL(8), parameter   :: A = 0.4
  REAL(8), parameter   :: delta  = 0.1
  REAL(8), parameter   :: tol  = 1.0e-7
  REAL(8) :: maxDiff  = 10.0
  REAL(8) :: kmin, kmax, rtrn, kss, yss, css


  REAL(8), DIMENSION(nkgrid) 	:: k_grid, kp_grid, V_old, V_new, k_pol, argvalue, argmax
  REAL(8), DIMENSION(nkgrid, nkgrid) :: temp


  ! Find Steady-State
  kss   = ((1 + beta * (delta - 1))/(alpha * beta * A))**(1 / (alpha - 1));
  yss   = A * kss**alpha + (1 - delta)*kss;
  css   = yss - kss;

  print *, 'Steady States are'
  print *, 'Output     : ', yss
  print *, 'Capital    : ', kss
  print *, 'Consumption: ', css
  print *

  ! Construct Capital GRID
  kmin = 0.5*kss; kmax = 1.5*kss;
  call linspace(kmin, kmax, nkgrid, k_grid);
  call linspace(kmin, kmax, nkgrid, kp_grid);

  DO WHILE (maxDiff > tol)
    iter = iter + 1;

    DO i = 1, nkgrid
        DO j = 1, nkgrid
          rtrn = A*(k_grid(i)**alpha) + (1-delta)*k_grid(i) - kp_grid(j);

          IF (rtrn < 0.0) THEN
            temp(i, j) = -100.0;
          ELSE
            temp(i, j) = log(rtrn) + beta*V_old(j);
          END IF

        END DO
    END DO

    ! Find Max and argmax
    ![argvalue, argmax]= max(temp,[],2);
    argvalue = maxval(temp, dim=2)
    V_new = argvalue;
    argmax = maxloc(temp, dim=2)

     DO i = 1, nkgrid
          k_pol(i) = k_grid(argmax(i));
     END DO

     maxDiff = maxval((abs(V_new - V_old)))
     V_old = V_new;
  END DO

  print *, 'Opitmal Value Function is: '
  print *, V_new
  print *
  print *, 'Opitmal Policy Function is: '
  print *, k_pol



END PROGRAM
